lava-api/src/job.rs

```diff
@@ -538,6 +539,48 @@ pub async fn cancel_job(lava: &Lava, id: i64) -> Result<(), CancellationError> {
     }
 }
 
+#[derive(Error, Debug)]
+pub enum ResultsError {
+    #[error("Request failed {0}")]
+    Request(#[from] reqwest::Error),
+    #[error("Unexpected reply: {0}")]
+    UnexpectedReply(reqwest::StatusCode),
+}
+
+pub fn job_results_as_junit(
+    lava: &Lava,
+    id: i64,
+) -> impl Stream<Item = Result<Bytes, ResultsError>> + Send + Unpin + '_ {
+    let mut url = lava.base.clone();
+    url.path_segments_mut()
+        .unwrap()
+        .pop_if_empty()
+        .push("jobs")
+        .push(&id.to_string())
+        .push("junit")
+        .push("");
+
+    Box::pin(
+        async move {
+            let b: Pin<Box<dyn Stream<Item = Result<Bytes, ResultsError>> + Send>> =
+                match lava.client.get(url).send().await {
+                    Ok(res) => match res.status() {
+                        StatusCode::OK => Box::pin(
+                            res.bytes_stream()
+                                .map(|item| item.map_err(ResultsError::from)),
+                        ),
+                        s => Box::pin(stream::once(async move {
+                            Err(ResultsError::UnexpectedReply(s))
+                        })),
+                    },
+                    Err(e) => Box::pin(stream::once(async move { Err(e.into()) })),
+                };
+            b
+        }
+        .flatten_stream(),
+    )
+}
```

Comment:
That's a heck of  lot of hoops to jump through just becuase `job_results_as_junit` is a sync method :) I've played this out a bit and the above can be collapsed into if the function is async:

```rust
    let res = lava.client.get(url).send().await?;
    match res.status() {
        StatusCode::OK => Ok(res.bytes_stream().map_err(ResultsError::from)),
        s => Err(ResultsError::UnexpectedReply(s)),
    }
```

Which is way more pleasant to read :)

Full patch in: https://github.com/collabora/lava-api/commit/a49922a69a8910d7f309b46a8592face4ffa1253

Reply:
I did start with basically what your patch shows, minus the neat `map_err` method from `TryStreamExt` that I was unaware of. I switched to the code above to simplify the callers.

As your patch shows, it's not really about sync vs async. The largest part of these changes is whether you want the errors to be rolled into the stream (simplifying the caller) or to have two different locations for errors to emerge at (immediately, or during stream reading). By rolling the errors in, we permit this to be used in quite naive ways in its callers (i.e. no additional error to map), we're purely returning the thing the caller wants. Alternatively, by separating the errors, sure we can make this code prettier.

Ultimately, it's a judgement call whether you pay the price in the library (once), or in the clients that need it, and while I generally still think this is right trade-off (putting this code here, making the library function interface simpler, at the cost of the details) there's no indisputably correct choice as far as I can see.

By making the change you suggest, we end up with pretty similar code elsewhere (in `lava-gitlab-runner` for one place) to roll the errors into the stream and undo the asyncness. It's not clear whether anywhere else will want to do the same thing, but I think it's arguable that having a pure stream as a return is generally nicer for future clients to work with.

If you're unmoved by this argument ;) then let me know, and I will apply the diffs I made when experimenting with this, and roll it back.

Reply:
For me there two quite different error topic anyway;  One is "we couldn't even start the stream' and the other is "something went wrong during streaming", which seems more then reasonable to split anyway in an API (Which matches what reqwest does as well!).

So yeah i don't see the point; I don't see it as a simplification of the API in the first place and it requires silly gymnastics to implement which isn't great for maintainability